cmd/az-log/cmd/common.go

```diff
@@ -0,0 +1,241 @@
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cmd
+
+import (
+	"bufio"
+	"context"
+	"fmt"
+	"log"
+	"os"
+	"path/filepath"
+	"regexp"
+	"strings"
+	"time"
+
+	homedir "github.com/mitchellh/go-homedir"
+	"github.com/spf13/cobra"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/clientcmd"
+	consts "sigs.k8s.io/azuredisk-csi-driver/pkg/azureconstants"
+)
+
+const (
+	AzureDiskContainer = "azuredisk"
+	RFC3339Format      = `^\d{4}-(\d{2})-(\d{2})T(\d{2}:\d{2}:\d{2}(.\d+)?)`
+	KlogTimeFormat     = `(\d{4}) (\d{2}:\d{2}:\d{2}(.\d+)?)`
+)
+
+func GetFlags(cmd *cobra.Command) ([]string, []string, []string, string, string, bool, bool) {
+	volumes, _ := cmd.Flags().GetStringSlice("volume")
+	nodes, _ := cmd.Flags().GetStringSlice("node")
+	requestIds, _ := cmd.Flags().GetStringSlice("request-id")
+	since, _ := cmd.Flags().GetString("since")
+	sinceTime, _ := cmd.Flags().GetString("since-time")
+	isFollow, _ := cmd.Flags().GetBool("follow")
+	isPrevious, _ := cmd.Flags().GetBool("previous")
+
+	if since != "" && sinceTime != "" {
+		fmt.Println("error: only one of --since/--since-time may be specified")
+		os.Exit(0)
+	}
+
+	return volumes, nodes, requestIds, since, sinceTime, isFollow, isPrevious
+}
+
+func getConfig() *rest.Config {
+	var kubeconfig string
+	if home, _ := homedir.Dir(); home != "" {
+		kubeconfig = filepath.Join(home, ".kube", "config")
+	}
+
+	config, err := clientcmd.BuildConfigFromFlags("", kubeconfig)
+	if err != nil {
+		panic(err.Error())
+	}
+
+	return config
+}
+
+func getKubernetesClientset(config *rest.Config) *kubernetes.Clientset {
+	clientsetK8s, err := kubernetes.NewForConfig(config)
+	if err != nil {
+		panic(err.Error())
+	}
+	return clientsetK8s
+}
+
+func GetLogsByAzDriverPod(clientsetK8s kubernetes.Interface, podName string, container string, volumes []string,
+	nodes []string, requestIds []string, since string, sinceTime string, isFollow bool, isPrevious bool) {
+
+	v1PodLogOptions := v1.PodLogOptions{
+		Container: container,
+		Follow:    isFollow,
+	}
+
+	// If since/sinceTime is specified, data type conversion and set up PodLogOptions
+	if since != "" {
+		d, err := time.ParseDuration(since)
+		if err != nil {
+			fmt.Println(err.Error())
+			os.Exit(0)
+		}
+		timeDuration := int64(d.Seconds())
+		v1PodLogOptions.SinceSeconds = &timeDuration
+	} else if sinceTime != "" {
+		t, err := TimestampFormatValidation(sinceTime)
+		if err != nil {
+			fmt.Println(err.Error())
+			os.Exit(0)
+		}
+		timestamp := metav1.NewTime(t)
+		v1PodLogOptions.SinceTime = &timestamp
+	}
+
+	podLogOptions := make([]v1.PodLogOptions, 0)
```

Comment:
Is there a need to have multiple PodLogOption objects in a slice? Can we just use one instead, with all the options set?

Reply:
When flag --previous is true, one PodLogOptions object is set field Previous = true to print logs from previous container and the other one is set Previous = false to reach logs from current container. That's why two objects with different field values are used.